packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -943,20 +925,47 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		}
 	}
 
-	public async getPendingBlobs(waitBlobsToAttach?: boolean): Promise<IPendingBlobs> {
-		if (waitBlobsToAttach) {
-			await this.stashPendingBlobs();
+	private async stashPendingBlobs(ids): Promise<void> {
+		for (const localId of ids) {
+			const entry = this.pendingBlobs.get(localId) as PendingBlob;
+			if (!entry.opsent) {
+				this.sendBlobAttachOp(localId, entry.storageId);
+			}
+			entry.handleP.resolve(this.getBlobHandle(localId));
 		}
+		return new Promise<void>((resolve) => {
+			if (this.allBlobsAttached) {
+				resolve();
+			} else {
+				this.once("allBlobsAttached", () => {
+					resolve();
+				});
+			}
+		});
+	}
+
+	public async getPendingBlobs(waitBlobsToAttach?: boolean): Promise<IPendingBlobs> {
 		const blobs = {};
-		for (const [key, entry] of this.pendingBlobs) {
-			blobs[key] = {
-				blob: bufferToString(entry.blob, "base64"),
-				storageId: entry.storageId,
-				attached: entry.attached,
-				acked: entry.acked,
-				minTTLInSeconds: entry.minTTLInSeconds,
-				uploadTime: entry.uploadTime,
-			};
+		const localIds = new Set<string>();
```

Comment:
why use ids, and not just the entries themselves, seems like it would simplify some things

Reply:
I believed using ids would be cleaner than the entries, given that they are ids. So instead of using pendingBlobs.keys (localId) do you want me to use pendingBlobs.values() (entry:PendingBlobs) ?

Reply:
i don't think the above is quite right, particularly around the await for all blobs attached. in both my code an your code i think there is a race condition where if a blob is created before the event fires, it won't get a handle, and this will block forever.